File: HW4/hw4.py
import sys

class readWrite:
	def __init__(self):
		self.temp = ""
		self.contents = []
		self.info = []
		
	def execute(self, file, tempFile):
		
		
		# https://www.guru99.com/reading-and-writing-files-in-python.html
		# https://stackoverflow.com/questions/9282967/how-to-open-a-file-using-the-open-with-statement
		with open(tempFile, 'r') as readTemplate:
			for content in readTemplate:
				self.temp = self.temp + content
		self.temp = self.temp.strip()
		delimiter = '\t' 
		
		with open(file, 'r') as readInput:
			contents = [line.strip() for line in readInput.readline().split(delimiter)]
			
			for input in readInput:
				swap = scoreData()
				data = [line.strip() for line in input.split(delimiter)]
				for i in range(len(contents)):
					swap.pushInfo(contents[i], data[i])
				
				self.info.append(swap)
		for data in self.info:
			data.fileOutput(self.temp)

class scoreData:
	def __init__(self):
		self.contents = {}
	
	def fileOutput(self, tempFile):
	# Generate a new file and write output
		for key in self.contents.keys():
			tempFile = tempFile.replace('<<'+key+'>>', self.contents[key])
			# Same idea as C#
			
		file = self.contents['ID']
		file = file + '.txt' #same as C# implementation
		with open(file, 'w') as writeOut:
			writeOut.write(tempFile)
		
	def pushInfo(self, location, data):
	# Add data to set to writeout 
		self.contents[location] = data

if __name__ == "__main__":
# https://stackoverflow.com/questions/419163/what-does-if-name-main-do

	runner = readWrite()
	# http://www.pythonforbeginners.com/system/python-sys-argv
	if (len(sys.argv) != 2):
		runner.execute(sys.argv[1], sys.argv[2])
	else:
		print("An incorrect number of files was used.")
		

File: HW4/hw4.cs
using System.Collections.Generic;
using System;
using System.IO;

public class hw4 {
	public static void Main(string[] args) {
		readWrite prog = new readWrite();
		int inputGiven = args.Length;
		
		// Safety check to make program more reliable
		if (inputGiven != 2) {
			Console.WriteLine("An incorrect number of files was used.");
		}
		else {
			prog.execute(args[0], args[1]);
		}
	}
}	

public class scoreData {
	public Dictionary<string, string> records;
	
	public scoreData() {
		records = new Dictionary<string,string>();
		// Used as reference: http://www.tutorialsteacher.com/csharp/csharp-dictionary
		//					  https://www.dotnetperls.com/dictionary
	}
	
	public void fileOutput(string tmpFile) {
		string copy = tmpFile;
		int total = 0;
		
		if (tmpFile != null) {
			int tagNumber = (tmpFile.Length - copy.Replace("<<", "").Length) / 2;
			// There is content in the file to write out
			foreach(string key in records.Keys) {
			// https://msdn.microsoft.com/en-us/library/fk49wtc1(v=vs.110).aspx
			if (total <= tagNumber) {
			tmpFile = tmpFile.Replace("<<"+key+">>", records[key]);
			total++;
			}
			}
		}
		
		// https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/file-system/how-to-write-to-a-text-file
		// Write the file name
		string file = records["ID"];
		file = file + ".txt"; // Add .txt at end
		using (StreamWriter output = File.CreateText(file)) {
			output.WriteLine(tmpFile);
		}
	}
}


public class readWrite {
	public List<string> columns;
	// https://stackoverflow.com/questions/12384905/string-split-function-in-c-sharp-tab-delimiter
	public List<scoreData> data = new List<scoreData>();
	public string temp; // Refactored output (sent in fileOutput method)
	
	public void execute(string file, string tmpFile) {

		// https://docs.microsoft.com/en-us/dotnet/standard/io/how-to-read-text-from-a-file
		using (StreamReader input = new StreamReader(tmpFile)) {
			string content; // Make global above??
			while ((content = input.ReadLine()) != null) {
				temp = temp + content + "\n"; 
			}
			if (content != null) {
				temp = temp.Trim();
			}
		}
		using (StreamReader input = new StreamReader(file)) {
			String content = input.ReadLine();
			columns = new List<string>(content.Split('\t'));
			// https://stackoverflow.com/questions/3575029/c-sharp-liststring-to-string-with-delimiter
			
			while((content = input.ReadLine()) != null) {
				List<string> columnData;
				columnData = new List<string>(content.Split('\t'));
				scoreData scores = new scoreData();
				
				int count = columns.Count;
				for (int i = 0; i < count; i++) {
					scores.records.Add(columns[i], columnData[i]);
				}
				data.Add(scores);
			}
		}
		foreach(scoreData score in data) {
			score.fileOutput(temp);
		}
	}
}
